File: app/src/main/java/com/lennydennis/zerakiapp/ui/viewmodels/RoomFragmentViewModel.java
package com.lennydennis.zerakiapp.ui.viewmodels;

import android.Manifest;
import android.app.Application;
import android.content.pm.PackageManager;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import com.lennydennis.zerakiapp.model.AccessTokenState;
import com.lennydennis.zerakiapp.repositories.AccessTokenRepo;
import com.lennydennis.zerakiapp.repositories.AccessTokenRepoImpl;
import com.twilio.video.CameraCapturer;

public class RoomFragmentViewModel extends AndroidViewModel {

    private AccessTokenRepo mAccessTokenRepo = new AccessTokenRepoImpl();
    public MutableLiveData<AccessTokenState> mAccessTokenMutableLiveData = new MutableLiveData<>();

    public RoomFragmentViewModel(@NonNull Application application) {
        super(application);
    }

    public boolean checkPermissionForCameraAndMicrophone() {
        int resultCamera = ContextCompat.checkSelfPermission(getApplication(), Manifest.permission.CAMERA);
        int resultMic = ContextCompat.checkSelfPermission(getApplication(), Manifest.permission.RECORD_AUDIO);
        return (resultCamera == PackageManager.PERMISSION_GRANTED) && (resultMic == PackageManager.PERMISSION_GRANTED);
    }

    public CameraCapturer.CameraSource getAvailableCameraSource() {
        return (CameraCapturer.isSourceAvailable(CameraCapturer.CameraSource.FRONT_CAMERA)) ?
                (CameraCapturer.CameraSource.FRONT_CAMERA) :
                (CameraCapturer.CameraSource.BACK_CAMERA);
    }

    public MutableLiveData<AccessTokenState> fetchAccessToken(String userName, String roomName){
        return  mAccessTokenMutableLiveData = mAccessTokenRepo.fetchAccessToken(userName,roomName);
    }

}

File: app/src/main/java/com/lennydennis/zerakiapp/ui/viewmodels/RoomFragmentViewModel.kt
package com.lennydennis.zerakiapp.ui.viewmodels

import androidx.lifecycle.*
import com.lennydennis.zerakiapp.model.AccessTokenState
import com.lennydennis.zerakiapp.model.PeerToPeerRoomState
import com.lennydennis.zerakiapp.repositories.AccessTokenRepo
import com.lennydennis.zerakiapp.repositories.AccessTokenRepoImpl
import com.lennydennis.zerakiapp.repositories.PeerToPeerRoomRepo
import com.lennydennis.zerakiapp.repositories.PeerToPeerRoomRepoImpl
import com.lennydennis.zerakiapp.ui.rooms.RoomEvent
import com.twilio.video.app.ui.room.RoomManager
import kotlinx.coroutines.launch

class RoomFragmentViewModel(private val roomManager: RoomManager) : ViewModel() {

    val roomEvents: LiveData<RoomEvent?> = roomManager.viewEvents

    private val mAccessTokenRepo: AccessTokenRepo = AccessTokenRepoImpl()
    lateinit var mAccessTokenMutableLiveData: LiveData<AccessTokenState>

    private val mPeerToPeerRoomRepo: PeerToPeerRoomRepo = PeerToPeerRoomRepoImpl()
    lateinit var mPeerToPeerRoomMutableLiveData: LiveData<PeerToPeerRoomState>

    fun fetchAccessToken(userName: String?, roomName: String?): MutableLiveData<AccessTokenState?>? {
        return mAccessTokenRepo.fetchAccessToken(userName, roomName).also { mAccessTokenMutableLiveData = it }
    }

    fun createPeerToPeerRoom(roomName: String?): MutableLiveData<PeerToPeerRoomState?>? {
        return mPeerToPeerRoomRepo.createPeerToPeerRoom(roomName).also { mPeerToPeerRoomMutableLiveData = it }
    }

    fun connectToRoom(
            roomName: String,
            accessToken: String) =
            viewModelScope.launch {
                roomManager.connectToRoom(
                        roomName,
                        accessToken)
            }

    fun disconnect() {
        roomManager.disconnect()
    }

    class RoomViewModelFactory(private val roomManager: RoomManager) : ViewModelProvider.Factory {

        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return RoomFragmentViewModel(roomManager) as T
        }
    }
}
